### PR TITLE
feat: mlflow: don't rebuild container images if content doesn't change

### PR DESCRIPTION
**Pull request by flaviodsr**
This PR adds the blueprint detailing the changes required to implement
such feature.

### CODE REVIEW

**Comment by jsuchome** [comment]
needs rebase

File: docs/blueprints/002-reuse-training-image.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+For checking if building the image is required, instead of tagging the image based on the codeset revision, tag it using a hash generated from the dependencies listed on the `conda.yaml` file. With that tag, it is possible to use a tool, such as [docker-ls](https://github.com/mayflower/docker-ls), to check if there is already an image with the same dependencies already present on the registry and build it only when such an image does not exist.
+
+### Affects
+
+1. **Example `mlflow-sklearn-e2e` workflow**
+   - The builder step should use the new mlflow-builder image and its output will be the image location from where it can be pulled, including the tag. This means that the
+     builder step will not rely on the `kaniko` and `builder-prep` tasks anymore.
```
Comment (jsuchome) [inline docs/blueprints/002-reuse-training-image.md:30]: Could you include the yaml snippet showing the change of the example workflow?

Comment (flaviodsr) [reply docs/blueprints/002-reuse-training-image.md:30]: done

**Comment by jsuchome** [comment]
So, do we still need our own kaniko and builder-prep tasks (as internal parts of tekton pipelines) when the builder actually uses kaniko by itself?

**Comment by flaviodsr** [comment]
> So, do we still need our own kaniko and builder-prep tasks (as internal parts of tekton pipelines) when the builder actually uses kaniko by itself?

For the example workflow it is not necessary anymore

**Comment by jsuchome** [comment]
> > So, do we still need our own kaniko and builder-prep tasks (as internal parts of tekton pipelines) when the builder actually uses kaniko by itself?
> 
> For the example workflow it is not necessary anymore

1. Could you write it explicitly somewhere in this blueprint?
2. Can you see a case where such steps would be still necessary?

**Comment by flaviodsr** [comment]
> > > So, do we still need our own kaniko and builder-prep tasks (as internal parts of tekton pipelines) when the builder actually uses kaniko by itself?
> > 
> > 
> > For the example workflow it is not necessary anymore
> 
> 1. Could you write it explicitly somewhere in this blueprint?

It is already there: https://github.com/fuseml/fuseml/pull/173/files#diff-f6c05f280879750242d8efddf153dec6ef3dd9c09530e34c557eed18a7793320R30

> 2. Can you see a case where such steps would be still necessary?

Yes, if someone does not want to deal with kaniko, he can still use the task to build the image as long as a Dockerfile is provided.

**Review by jsuchome — APPROVED** [review]

**Pull request merged**